### DISCUSSION TITLE
Function executed finish reason

### DISCUSSION THREAD

**Discussion by simetelum** [KeyserDSoze/Rystem.OpenAi#47] [Ideas]
Hi, would it be possible to add another FinishReason when function is auto executed. For example, if function is NOT auto executed, finish reason would be "null", if function is auto executed, finish reason is "stop" - the same as ordinary chat response. Would it be possible to add `FunctionExecutedFinishReason = "functionExecuted"` or something like that and replace it with "stop" if function is auto executed so the caller knows if function is executed when functions are registered using DI?

**Comment by KeyserDSoze** [comment]
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/1e9260da-5205-4ffd-919e-05ce286d3878)
 Something like this?

**Reply by simetelum** [reply]
Yes, that would be perfect

**Reply by KeyserDSoze** [reply]
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/4ef684d0-0407-454c-b02a-50ac07077805)

**Reply by KeyserDSoze** [reply]
In the end I opted for this solution
![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/40713438/4658aaca-a3b5-4e9e-93d5-4fd7d364ec34)

Auto when you use auto, and when you inject the function message by yourself instead you may find the same string without auto.

**Reply by simetelum** [reply]
Ok, that works for me, but I don't understand this: 

> and when you inject the function message by yourself instead

Could you explain what do you mean by injecting function message by myself?

**Reply by KeyserDSoze** [reply]
request
        .AddFunctionMessage(functionName, "{\"temperature\": \"22\", \"unit\": \"celsius\", \"description\": \"Sunny\"}");

and with false or default in Execute...Async();

            response = await request
                .ExecuteAndCalculateCostAsync();

For instance I tested the solution with this integration test
```
var openAiApi = _openAiFactory.Create(name);
            Assert.NotNull(openAiApi.Chat);
            var functionName = "get_current_weather";
            var request = openAiApi.Chat
                .RequestWithUserMessage("What is the weather like in Boston?")
                .WithModel(ChatModelType.Gpt35Turbo)
                .WithFunction(new JsonFunction
                {
                    Name = functionName,
                    Description = "Get the current weather in a given location",
                    Parameters = new JsonFunctionNonPrimitiveProperty()
                        .AddPrimitive("location", new JsonFunctionProperty
                        {
                            Type = "string",
                            Description = "The city and state, e.g. San Francisco, CA"
                        })
                        .AddEnum("unit", new JsonFunctionEnumProperty
                        {
                            Type = "string",
                            Enums = new List<string> { "celsius", "fahrenheit" }
                        })
                        .AddRequired("location")
                });

            var response = await request
                .ExecuteAndCalculateCostAsync();

            var function = response.Result.Choices[0].Message.Function;
            Assert.NotNull(function);
            Assert.Equal(function.Name, functionName);
            var weatherRequest = JsonSerializer.Deserialize<WeatherRequest>(function.Arguments);
            Assert.NotNull(weatherRequest?.Location);

            request
                .AddFunctionMessage(functionName, "{\"temperature\": \"22\", \"unit\": \"celsius\", \"description\": \"Sunny\"}");
            response = await request
                .ExecuteAndCalculateCostAsync();

            var content = response.Result.Choices[0].Message.Content;
            Assert.Equal("functionExecuted", response.Result.Choices[0].FinishReason);
            Assert.NotNull(content);
```

**Reply by simetelum** [reply]
Ok, I get it, I do not use it that way, but if I do, I would definitely need  both of those FunctionFinishReason. 

Thank you.

**Reply by KeyserDSoze** [reply]
it's now online the new version 3.3.3 with this feature.
